File: sdk/sdk-js/src/methods/apikey.ts
import { ApiKeyInfo, ApiKeyEnumerationQuery, CredentialCreateResponse, EnumerationResult } from '../models';
import { HttpClient } from '../http-client';
import { NetLedgerValidationError } from '../errors';

/**
 * API key management operations.
 */
export class ApiKeyMethods {
    private readonly client: HttpClient;

    constructor(client: HttpClient) {
        this.client = client;
    }

    /**
     * Create a new API key.
     * @param name Display name for the key.
     * @param isAdmin Whether the key has admin privileges.
     * @returns The created credential and one-time secret key.
     */
    async create(name: string, isAdmin: boolean = false): Promise<CredentialCreateResponse> {
        if (!name || name.trim() === '') {
            throw new NetLedgerValidationError('API key name cannot be empty', 'name');
        }
        const response = await this.client.put<CredentialCreateResponse>(
            '/v1/credentials',
            { Name: name, IsAdmin: isAdmin }
        );
        if (!response.Data) {
            throw new Error('No data returned from server');
        }
        return response.Data;
    }

    /**
     * Enumerate API keys.
     * @param query Query parameters.
     * @returns Enumeration result (key values not included).
     */
    async enumerate(query?: ApiKeyEnumerationQuery): Promise<EnumerationResult<ApiKeyInfo>> {
        const params = new URLSearchParams();
        if (query) {
            if (query.MaxResults !== undefined) params.append('maxResults', query.MaxResults.toString());
            if (query.Skip !== undefined) params.append('skip', query.Skip.toString());
            if (query.TenantId !== undefined) params.append('tenantId', query.TenantId);
        }
        const queryString = params.toString();
        const path = queryString ? `/v1/credentials?${queryString}` : '/v1/credentials';
        const response = await this.client.get<EnumerationResult<ApiKeyInfo>>(path);
        return response.Data || { TotalRecords: 0, RecordsRemaining: 0, EndOfResults: true };
    }

    /**
     * Revoke (delete) an API key.
     * @param apiKeyGuid The API key GUID.
     */
    async revoke(apiKeyGuid: string): Promise<void> {
        await this.client.delete(`/v1/credentials/${apiKeyGuid}`);
    }
}

File: sdk/sdk-python/netledger_sdk/methods/apikey.py
"""API key management operations for the NetLedger API."""

from typing import Optional

from ..http_client import HttpClient
from ..models import ApiKeyInfo, ApiKeyEnumerationQuery, EnumerationResult
from ..exceptions import NetLedgerValidationError


class ApiKeyMethods:
    """API key management operations."""

    def __init__(self, client: HttpClient):
        """
        Initialize API key methods.

        Args:
            client: The HTTP client to use.
        """
        self._client = client

    def create(self, name: str, is_admin: bool = False) -> ApiKeyInfo:
        """
        Create a new API key.

        Args:
            name: Display name for the key.
            is_admin: Whether the key has admin privileges.

        Returns:
            The created API key info (includes the key value).

        Raises:
            NetLedgerValidationError: If the name is empty.
            NetLedgerConnectionError: If unable to connect to the server.
            NetLedgerApiError: If the server returns an error (401 if not authorized).
        """
        if not name or not name.strip():
            raise NetLedgerValidationError('API key name cannot be empty', 'name')

        body = {'name': name, 'isAdmin': is_admin}
        response = self._client.put('/v1/apikeys', body)
        if not response.data:
            raise ValueError('No data returned from server')
        return ApiKeyInfo.from_dict(response.data)

    def enumerate(self, query: Optional[ApiKeyEnumerationQuery] = None) -> EnumerationResult:
        """
        Enumerate API keys.

        Args:
            query: Query parameters.

        Returns:
            Enumeration result (key values not included).

        Raises:
            NetLedgerConnectionError: If unable to connect to the server.
            NetLedgerApiError: If the server returns an error (401 if not authorized).
        """
        if query is None:
            query = ApiKeyEnumerationQuery()

        path = f'/v1/apikeys?maxResults={query.max_results}&skip={query.skip}'
        response = self._client.get(path)

        if not response.data:
            return EnumerationResult()

        return EnumerationResult.from_dict(response.data, ApiKeyInfo.from_dict)

    def revoke(self, api_key_guid: str) -> None:
        """
        Revoke (delete) an API key.

        Args:
            api_key_guid: The API key GUID.

        Raises:
            NetLedgerConnectionError: If unable to connect to the server.
            NetLedgerApiError: If the server returns an error (401 if not authorized, 404 if not found).
        """
        self._client.delete(f'/v1/apikeys/{api_key_guid}')
